wicketd/src/update_tracker.rs

```diff
@@ -380,31 +382,158 @@ impl UpdateDriver {
+                    artifact: plan.trampoline_phase_1.id.clone(),
+                    reason: format!("{err:#}"),
+                }
+            })?;
+
+        self.wait_for_first_installinator_progress(
+            plan,
+            uploaded_trampoline_phase2_id,
+        )
+        .await
+        .map_err(|err| {
             UpdateEventFailureKind::ArtifactUpdateFailed {
-                // `trampoline_phase_1` and `trampoline_phase_2` have the same
-                // artifact ID, so the choice here is arbitrary.
+                // `trampoline_phase_1` and `trampoline_phase_2` have the
+                // same artifact ID, so the choice here is arbitrary.
                 artifact: plan.trampoline_phase_1.id.clone(),
                 reason: format!("{err:#}"),
             }
         })?;
 
-        // TODO wait for installinator completion, ideally with progress for
-        // both:
-        //
-        // a) fetching the the installinator image from MGS
-        // b) installinator progress
+        // TODO wait for installinator completion, ideally with progress along
+        // the way.
 
         // TODO power down host and deliver matching host phase 1
         _ = plan.host_phase_1;
 
         Ok(())
     }
 
+    async fn wait_for_first_installinator_progress(
+        &self,
+        plan: &UpdatePlan,
+        uploaded_trampoline_phase2_id: HostPhase2RecoveryImageId,
+    ) -> anyhow::Result<()> {
+        const MGS_PROGRESS_POLL_INTERVAL: Duration = Duration::from_secs(3);
+
+        // Waiting for the installinator to start is a little strange. It can't
+        // start until the host boots, which requires all the normal boot things
+        // (DRAM training, etc.), but also fetching the trampoline phase 2 image
+        // over the management network -> SP -> uart path, which runs at about
+        // 167 KiB/sec. This is a _long_ time to wait with no visible progress,
+        // so we'll query MGS for progress of that phase 2 trampoline delivery.
+        // However, this query is "best effort" - MGS is observing progress
+        // indirectly (i.e., "what was the last request for a phase 2 image I
+        // got from this SP"), and it isn't definitive. We'll still report it as
+        // long as it matches the trampoline image we're expecting the SP to be
+        // pulling, but it's possible we could be seeing stale status from a
+        // previous update attempt with the same image.
+        //
+        // To start, _clear out_ the most recent status that MGS may have
+        // cached, so we don't see any stale progress from a previous update
+        // through this SP. If somehow we've lost the race and our SP is already
+        // actively requesting host blocks, this will discard a real progress
+        // message, but that's fine - in that case we expect to see another real
+        // one imminently. It's possible (but hopefully unlikely?) that the SP
```

Comment:
It would be nice to eliminate this possibility entirely, and during MUP we likely will. I'm not too concerned right now though.

I really appreciate the detail in this comment. It's going to help later when I forget what I just reviewed :)